### DISCUSSION TITLE
class 'AttributeError'

### DISCUSSION THREAD

**Discussion by leandermarlon** [netbox-community/netbox#16105] [Help Wanted!]
Hi

I have Netbox on Docker, and since I upgraded to a newer Netbox version, every time I try to access the Device Tab, I get this error:
![image](https://github.com/netbox-community/netbox/assets/169671019/d07a392f-3e3d-475b-9c51-c88b95c1aa56)

Can somone assist, please?

Thank you!

**Comment by micko** [comment]
Having the same issue with 4.1 right after upgrade from 4.0.8

**Comment by candlerb** [comment]
Please obtain a full backtrace by setting DEBUG=True in the environment (this is picked up by [configuration.py](https://github.com/netbox-community/netbox-docker/blob/release/configuration/configuration.py#L165)) and post it here.

**Reply by micko** [reply]
Here is the output:

`Environment:


Request Method: GET
Request URL: https://netbox.silverairways.com/dcim/devices/

Django Version: 5.0.9
Python Version: 3.11.2
Installed Applications:
['django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.humanize',
 'django.forms',
 'corsheaders',
 'debug_toolbar',
 'django_filters',
 'django_htmx',
 'django_tables2',
 'django_prometheus',
 'strawberry_django',
 'mptt',
 'rest_framework',
 'social_django',
 'taggit',
 'timezone_field',
 'core',
 'account',
 'circuits',
 'dcim',
 'ipam',
 'extras',
 'tenancy',
 'users',
 'utilities',
 'virtualization',
 'vpn',
 'wireless',
 'django_rq',
 'drf_spectacular',
 'drf_spectacular_sidecar']
Installed Middleware:
['strawberry_django.middlewares.debug_toolbar.DebugToolbarMiddleware',
 'corsheaders.middleware.CorsMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.locale.LocaleMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'django_htmx.middleware.HtmxMiddleware',
 'netbox.middleware.RemoteUserMiddleware',
 'netbox.middleware.CoreMiddleware',
 'netbox.middleware.MaintenanceModeMiddleware']



Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/views/generic/base.py", line 77, in dispatch
    return super().dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/utilities/views.py", line 125, in dispatch
    return super().dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/utilities/views.py", line 39, in dispatch
    return super().dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/views/generic/bulk_views.py", line 190, in get
    'filter_form': self.filterset_form(request.GET) if self.filterset_form else None,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/forms/base.py", line 173, in __init__
    super().__init__(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/forms/mixins.py", line 31, in __init__
    self._append_customfield_fields()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/forms/mixins.py", line 57, in _append_customfield_fields
    self.fields[field_name] = self._get_form_field(customfield)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/forms/base.py", line 188, in _get_form_field
    return customfield.to_form_field(set_initial=False, enforce_required=False, enforce_visibility=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/extras/models/customfields.py", line 529, in to_form_field
    queryset=model.objects.all(),
             ^^^^^^^^^^^^^

Exception Type: AttributeError at /dcim/devices/
Exception Value: 'NoneType' object has no attribute 'objects'
`

**Reply by candlerb** [reply]
It looks to be something to do with a custom field, which is of type "Object", but has no object class configured.

Relevant code from `/opt/netbox/netbox/extras/models/customfields.py`:

```
    def to_form_field(self, set_initial=True, enforce_required=True, enforce_visibility=True, for_csv_import=False):
...
        elif self.type == CustomFieldTypeChoices.TYPE_OBJECT:
            model = self.related_object_type.model_class()    # <<< returns None
            field_class = CSVModelChoiceField if for_csv_import else DynamicModelChoiceField
            field = field_class(
                queryset=model.objects.all(),    # <<< failing here
                required=required,
                initial=initial,
                query_params=self.related_object_filter
            )
```

Is it possible that you had in the past added a custom field to a plugin model, and the plugin no longer exists?

If you can navigate to `/extras/custom-fields/` maybe you can find and fix the offending custom field there. If not, try this ugly hack (untested) to the code shown above, to see if it gets you working to the point where you can edit the custom fields:

```
-            model = self.related_object_type.model_class()
+            from dcim.models import Site
+            model = self.related_object_type.model_class() or Site
```

Or try a SQL query:

```
select * from extras_customfield where related_object_type_id is null and type='object';
```

I wouldn't just delete the row via SQL, because that won't fix up any custom_field_data on objects, but you could try changing it to a different type, such as `'text'`:

```
update extras_customfield set type='text' where related_object_type_id is null and type='object';
```

Make sure you [backup](https://netboxlabs.com/docs/netbox/en/stable/administration/replicating-netbox/) your database first before trying any SQL surgery.

In any case, it would seem to be a good idea if `model = self.related_object_type.model_class()` were to check for a null result and handle it sensibly (for both object and multiobject field types), even though the `CustomField.clean()` method *should* prevent such invalid custom fields being created.

**Reply by micko** [reply]
The issue had to do with one `extra` plugin (which had added the extra fields) that was not ready for 4.1 and I had to remove it.
